File: 2020/golang/day2.go
package main

import (
	"log"
	"strings"
)

// from utils import getInputData, getSpiltList

// if __name__ == "__main__":
//     data = getInputData(year=2020, day=2)

//     data = getSpiltList(data=data, splitChar=":")

//     rangeVals = []
//     charVals = []
//     password_vals = []

//     for line in data:
//         rangeVals.append(list(map(int, line[0][0:-2].split("-"))))
//         charVals.append(line[0][-1])
//         password_vals.append(line[1][1:])

//     # Part 1

//     valid = 0
//     for i in range(1000):
//         count = 0
//         for char in password_vals[i]:
//             if char == charVals[i]:
//                 count += 1

//         if count >= rangeVals[i][0] and count <= rangeVals[i][1]:
//             valid += 1

//     print(valid)

//     # Part 2
//     valid = 0
//     for i in range(1000):
//         lower_char = password_vals[i][(rangeVals[i][0] - 1)]
//         higher_char = password_vals[i][(rangeVals[i][1] - 1)]
//         if charVals[i] == lower_char or charVals[i] == higher_char:
//             if lower_char != higher_char:
//                 valid += 1

//     print(valid)

func day2() {

	data := getSplitList(getInputData(2020, 2), ":")

	day2part1(data)
	day2part2(data)

}

func day2part1(data [][]string) {

	var valid int

	for _, entry := range data {

		charPolicy := strings.Split(entry[0], " ")
		charLimits := StringSliceToIntegerSlice(strings.Split(charPolicy[0], "-"))
		char := strings.Trim(charPolicy[1], " ")
		password := strings.Trim(entry[1], " ")

		if count := strings.Count(password, char); count >= charLimits[0] && count <= charLimits[1] {
			valid += 1
		}
	}

	log.Println("Answer for Day 2, Part 1 is", valid)

}

func day2part2(data [][]string) {

	var valid int

	for _, entry := range data {

		charPolicy := strings.Split(entry[0], " ")
		charIndices := StringSliceToIntegerSlice(strings.Split(charPolicy[0], "-"))
		char := strings.Trim(charPolicy[1], " ")
		password := strings.Trim(entry[1], " ")

		isCharPresent := string(password[charIndices[0]-1]) == char || string(password[charIndices[1]-1]) == char
		isDifferent := password[charIndices[0]-1] != password[charIndices[1]-1]

		if isCharPresent && isDifferent {
			valid += 1
		}
	}

	log.Println("Answer for Day 2, Part 2 is", valid)

}

File: 2022/python/day2.py
from utils import getInputData

data = getInputData(year=2022, day=2)

parsed_data = []

for item in data:
    parsed_data.append(item.split())

sign = {'X': 1, 'Y':2, 'Z': 3}

total_score = 0

for combo in parsed_data:
    if (combo[0] == 'A' and combo[1] == 'Y') or (combo[0] == 'B' and combo[1] == 'Z') or (combo[0] == 'C' and combo[1] == 'X'):
        total_score += 6
        total_score += sign[combo[1]]
    elif (combo[0] == 'A' and combo[1] == 'Z') or (combo[0] == 'B' and combo[1] == 'X') or (combo[0] == 'C' and combo[1] == 'Y'):
        total_score += sign[combo[1]]
    else:
        total_score += 3
        total_score += sign[combo[1]]

print("Answer for part 1", total_score)

total_score = 0

for combo in parsed_data:
    if combo[1] == 'Z':
        total_score += 6
        if combo[0] == 'A':
            total_score += 2 # Opponent chose rock, we chose paper, we win.
        elif combo[0] == 'B':
            total_score += 3
        elif combo[0] == 'C':
            total_score += 1

    elif combo[1] == 'Y':
        total_score += 3
        if combo[0] == 'A':
            total_score += 1
        elif combo[0] == 'B':
            total_score += 2
        elif combo[0] == 'C':
            total_score += 3
    elif combo[1] == 'X':
        if combo[0] == 'A':
            total_score += 3 # Opponent chose rock, we chose scissor, we lose.
        elif combo[0] == 'B':
            total_score += 1
        elif combo[0] == 'C':
            total_score += 2

print("Answer for part 2", total_score)

        
